pkg/ingester/ingester_v2.go

```diff
@@ -738,13 +722,17 @@ func (i *Ingester) v2Push(ctx context.Context, req *cortexpb.WriteRequest) (*cor
 	)
 
 	// Walk the samples, appending them to the users database
-	app := db.Appender(ctx)
+	app := db.Appender(ctx).(extendedAppender)
 	for _, ts := range req.Timeseries {
-		// Check if we already have a cached reference for this series. Be aware
-		// that even if we have a reference it's not guaranteed to be still valid.
+		// Keeps a reference to labels copy, if it was needed. This is to avoid making a copy twice,
+		// once for TSDB, and second time for activeSeries map.
+		var copiedLabels []labels.Label
+
 		// The labels must be sorted (in our case, it's guaranteed a write request
 		// has sorted labels once hit the ingester).
-		cachedRef, copiedLabels, cachedRefExists := db.refCache.Ref(startAppend, cortexpb.FromLabelAdaptersToLabels(ts.Labels))
+
+		// Look up a reference for this series. Holding the appendLock ensures that no compaction will happen while we use it.
```

Comment:
> Holding the appendLock ensures that no compaction will happen while we use it.

Are you talking about `acquireAppendLock()` we have in Cortex ingester? It just ensures the forced compaction will not happen while there are on-going appends, but not the regular head compaction (see the `userDB.Compact()` call in `compactBlocks()`).

Reply:
What if TSDB `GetRef()` would also returned interned series labels so we'll just re-pass them to the `Append()`? It would also allow us to reuse the same interned strings in the active series tracking.

Reply:
I made a new version with `GetRef()` returning labels.